# **Практика 2. Фильтрация и сортировка данных в DataFrame**

[SmartLMS](https://edu.hse.ru/course/section.php?id=1221680)

Вы работаете с датасетом который содержит информацию по рейтингу раменов: `ramen.csv`

| **Признак** | **Описание**                             |
|-------------|------------------------------------------|
| brand       | бренд рамена                             |
| variety     | название рамена                          |
| style       | тип упаковки                             |
| country     | страна производства                      |
| stars       | средний рейтинг (от 0 до 5)              |
| votes       | количество людей, поставивших оценку     |


In [ ]:
import pandas as pd
df = pd.read_csv('data/ramen.csv')

## **Задание 1**

>*Определите тип данных каждого признака:*

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3180 entries, 0 to 3179
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   brand    3180 non-null   object 
 1   variety  3180 non-null   object 
 2   style    3178 non-null   object 
 3   country  3180 non-null   object 
 4   stars    3166 non-null   float64
 5   votes    3180 non-null   int64  
dtypes: float64(1), int64(1), object(4)
memory usage: 149.2+ KB


| **Признак** | **Тип данных**               |
|-------------|------------------------------|
| brand       | категориальный номинальный   |
| variety     | категориальный номинальный   |
| style       | категориальный номинальный   |
| country     | категориальный номинальный   |
| stars       | количественный непрерывный   |
| votes       | количественный дискретный    |


---

## **Задание 2**

>*Укажите количество раменов бренда **Nissin** (`brand`).*


In [6]:
cond = df['brand'] == 'Nissin'
df[cond].shape[0]

443

In [7]:
(df.brand == 'Nissin').sum()

np.int64(443)

**Ответ**: 443

---

## **Задание 3**

>*Укажите количество раменов c рейтингом более 3.5 (`stars`).*

In [10]:
cond = df['stars'] > 3.5
df[cond].shape[0]

1841

In [11]:
(df.stars > 3.5).sum()

np.int64(1841)

**Ответ**: 1841

---

## **Задание 4**

>*Укажите количество раменов в типе упаковки не **Bowl** (`style`) c рейтингом не более 4 (`stars`).*

In [14]:
cond = (df['style'] != 'Bowl') & (df['stars'] <= 4)
df[cond].shape[0]

1771

**Ответ**: 1771

---

## **Задание 5**

>*Укажите общее количество раменов бренда **Maruchan** (`brand`) или тех, у которых тип упаковки **Cup** (`style`).*


In [15]:
cond = (df['brand'] == 'Maruchan') | (df['style'] == 'Cup')
df[cond].shape[0]

639

**Ответ**: 639

---

## **Задание 6**

>*Укажите количество раменов бренда **Myojo** и **Samyang Foods** (`brand`).*

In [18]:
cond = (df['brand'] == 'Myojo') | (df['brand'] == 'Samyang Foods')
df[cond].shape[0]

172

In [19]:
cond = df['brand'].isin(['Myojo', 'Samyang Foods'])
df[cond].shape[0]

172

**Ответ**: 172

---

## **Задание 7**

>*Отсортируйте названия (`variety`) всех раменов по алфавиту. В ответ укажите индекс рамена, который будет выведен первым.*

In [20]:
df.sort_values('variety').head(1)

brand                        variety style country  stars  votes
2570  Wei Wei  "A" Series Artificial Chicken  Pack  Taiwan   3.25   2045

In [21]:
df.sort_values('variety').index[0]

np.int64(2570)

**Ответ**: 2570

---

## **Задание 8**

>*Найдите рамен, оценку которому поставило наибольшее количество людей (`votes`). В ответ укажите его название (`variety`).*

In [22]:
df.sort_values('votes', ascending=False).iloc[0].loc['variety']

'Demae Ramen Straight Noodle Spicy Tonkotsu Flavour Instant Noodle'

---

## **Задание 9**

>*Среди раменов, произведенных в Тайланде (**Thailand**; `country`), найдите рамен с наименьшим рейтингом (`stars`). В ответ укажите его название (`variety`).*

In [23]:
df[df['country']=='Thailand'].sort_values('stars').head(1)

brand                variety style   country  stars  votes
3052  Mama  Artificial PA-LO Duck  Pack  Thailand    0.1   1649

---

## **Задание 10**

>*Среди раменов бренда **Nissin** (`brand`), чей рейтинг равен 5 (`stars`). Найдите рамен, который оценило наименьшее количество людей (`votes`). В ответ укажите страну его производства (`country`).*


In [24]:
df[
    (df['brand']=='Nissin') & (df['stars']==5.0)
].sort_values('votes').head(1)

brand                      variety style country  stars  votes
1078  Nissin  Cup Noodles Singapore Laksa   Cup   Japan    5.0   1006

---

## **Задание 11**

>*Среди раменов произведенных в Японии (**Japan**; `country`) и тех чей рейтинг не равен 5 (`stars`). Найдите рамен, который оценило наибольшее количество людей (`votes`). В ответ укажите тип его упаковки (`style`).*


In [25]:
df[
    (df['country']=='Japan') & (df['stars']!=5.0)
].sort_values('votes').tail(1)

brand            variety style country  stars  votes
1417  Ishimaru  Tuna Flavor Ramen  Pack   Japan    4.5   3212

---

## **Задание 12**

>*Среди раменов бренда **Wai Wai** и **Vifon** (`brand`). Найдите рамен, с наибольшим рейтингом (`stars`). В ответ укажите его название (`variety`).*


In [26]:
df[
    df['brand'].isin(['Wai Wai', 'Vifon'])
].sort_values('stars').tail(1).iloc[0, 1]

'Viet Cuisine Bun Rieu Cua Sour Crab Soup Instant Rice Vermicelli'